rascal-textmate-core/src/main/rascal/lang/textmate/conversiontests/Walkthrough.rsc

```diff
@@ -0,0 +1,290 @@
+
+lexical Key    = Alnum+ !>> [a-z A-Z 0-9];
+lexical Number = @category="constant.numeric" Digit+ !>> [0-9];
+lexical String = @category="string.quoted.double" "\"" Alnum* "\"";
+
+// Basically, the conversion algorithm converts each Rascal non-terminal that is
+// *suitable for conversion* to a TextMate rule. For instance, `Number` is
+// converted to the following match pattern (in JSON):
+//
+// ```
+// {
+//   "match": "([\\u0030-\\u0039]+?(?![\\u0030-\\u0039]))",
+//   "name": "prod(lex(\"Number\"),[conditional(iter(lex(\"Digit\")),{\\not-follow(\\char-class([range(48,57)]))})],{tag(\"category\"(\"constant.numeric\"))})", 
+//   "captures": {
+//     "1": {
+//       "name": "constant.numeric"
+//     }
+//   }
+// }
+// ```
+//
+// Note: The regular expression (`match` property) is written in Oniguruma
+// format (following the TextMate grammar specification), using code units
+// instead of alphanumericals.
+//
+// Note: The name (`name` property) could be anything, but to simplify
+// debugging, the conversion algorithm uses (part of) the internal
+// representation of the Rascal non-terminal as the name.
+//
+// In general, a Rascal non-terminal is "suitable for conversion" when it
+// satisfies each of the following conditions:
+//
+//  1. It is non-recursive. (Recursion is prohibitively hard to faithfully
+//     convert; tail-recursion could be workable, but currently out of scope.)
+//
+//  2. It does not match newlines. (TextMate rules that involve matching against
+//     newlines are dysfunctional.)
+//
+//  3. It does not match the empty word. (TextMate rules that involve matching
+//     against the empty word are dysfunctional.)
+//
+//  4. It has a `@category` tag.
+//
+// For instance, `Number` and `String` are suitable for conversion, but `Value`
+// (violation of conditions 1 and 4), `Map` (violation of condition 1), and
+// `Key` (violation of condition 4) are not suitable.
+
+
+
+// ## Extension 1: Conversion of productions instead of non-terminals
+//
+// The first extension (layout) of the grammar looks like this:
+
+lexical Comment
+    = @category="comment.line.double-slash" line:  "//" (Alnum | Blank)* $
+    | @category="comment.block"             block: "/*" (Alnum | Space)* "*/";
+
+layout Layout = (Comment | Space)* !>> "//" !>> [\ \t\n];
+
+// `Comment` is *not* suitable for conversion, as it violates condition 2: the
+// corresponding TextMate rule would involve matching against newlines. However,
+// the matching against newlines is needed only for production `block`; not for
+// production `line`. Thus, conversion at the granularity of Rascal
+// non-terminals is actually too coarse.
+// 
+// To solve this issue, the conversion algorithm works at the granularity of
+// individual productions (specifically, `prod` constructors). For instance,
+// production `line` of `Comment` is individually converted to the following
+// match pattern, independently of production `block` (which is ignored):
```

Comment:
would it be possible to generate begin & end style textmate recognizers for `block`?